Adding extension for BAM

katallaxie opened the pull request:
Fixes #

BAM is a method to track the progress of your business processes via the occurrence of business events in your organization. Famously, many companies used BizTalk to implement BAM in their organization, but with open standards like CloudEvents this is changing. This extension is necessary, as the payload of an event is processed by primary systems, like payment or ERP and the payload of these events should not change; but the progress should be tracked in an BAM system for business users to observe the processing and to control these transactions.

## Proposed Changes

- Add extension with attributes that support the implementation or use of a BAM system in an enterprise.

**Release Note**

<!--
If this change has user-visible impact, write a release note in the block
below. If this change has no user-visible impact, no release note is needed.
-->

```release-note
Adds the business activity monitoring (BAM) known extension
```

duglin commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,96 @@
+
+## Attributes
+
+### businessTransactionId
```
We require attribute names to be all lowercase since they can appear in case insensitive locations, like http headers

duglin commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,96 @@
+- Constraints
+  - REQUIRED
+  - MUST be a non-empty string
+  - RECOMMENDED as monotonically increasing and contiguous identifier that is lexicographically-sortable. 
```
Can you wrap all lines at 80 columns?

duglin commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,96 @@
+- Description: The status of the business process transaction step associated with the event.
+- Constraints
+  - OPTIONAL
+  - MUST be a non-empty string
```
minor thing, add "if present, " to the start of lines like this where the attribute is OPTIONAL

duglin commented:
All, please review before tomorrow's call

duglin commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,117 @@
+
+## References
+
+- [Gartner Business Activity Monitoring](https://www.gartner.com/en/information-technology/glossary/business-activity-monitoring-bam)
```
I get a 404 for this

katallaxie replied:
:) it so happend that they changed their content structure. They started to include the language in their urls.

duglin commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,117 @@
+  - RECOMMENDED as monotonically increasing and contiguous identifier
+    that is lexicographically-sortable.
+
+### businessprocessid
```
Is there a doc we can point to that explains the relationship of the various entities described here?
For example, do process-transactions live within processes which live within transactions? So it's a M:N:1 relationship ?

katallaxie replied:
I will provide an brief explanation to those fields

duglin commented:
We discussed this on yesterday's call and overall there's no concern with add it, however @JemDay was wondering about the length of some of those attributes. Would it be possible to use `bam` in there to shorten them?

Just tossing this out there:
```
bamtxid
bampid
bamptxid
bamptxpid
bamptxstatus
bamptxcompleted
```

Does there need to be a `bamtxcompleted` too?

katallaxie commented:
@duglin great suggestion.

- Updated the naming of the fields.
- Provided a more in-depth explanation of how this works (like a state machine of the business process that is controlled by the events).
- Fixed the link (they change the URL schema).

markpeek commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,129 @@
+
+## Attributes
+
+### bamtxid
```
This is shortening the name further than I had thought on the call. As such, would it make sense to have it expanded for readability such as:
```
### bamtxid (Business Transaction ID)
```
Note: neither the spec of other extensions have this but those attributes are more readable. @duglin thoughts?

katallaxie replied:
I think this makes sense.

```markdown
### `bamtxid` (Business Transaction ID)
```

Would strike a balance between efficiency of the fields and a good documentation. I think the convention on not using CamelCase, for the obvious reason of compatibility between platforms, this strikes the balance of actually make the long intention readable by shortening it. But on the other hand make it clear in the docs what is should be used for.

duglin replied:
I agree that some of the names I came up with are a bit short and I'd be ok with making them a bit longer - I'm just not that creative :-). But if folks are ok with these short names then yes I agree with adding the descriptive long names in parens.

markpeek commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,129 @@
+
+## Usage
+
+When this extension is used, producers MUST set the value of
```
This should be changed to:
```
When this extension is used, producers MUST set the value of
the `bamtxid`, `bampid`,
`bamptxid`, and `bamptxsid`
attributes to the unique identifiers of the business process, transaction,
and transaction step associated with the event.

Intermediaries MUST NOT change the value of the `bamtxid`,
`bampid`, `bamptxid`,
and `bamptxsid` attributes.
```
since those other attribute names are no longer specified above. Again, I could see expanding the attribute name in parens for readability.

duglin commented on cloudevents/extensions/bam.md:
```diff
@@ -0,0 +1,127 @@
+
+## Attributes
+
+### bamtxid (Business Transaction ID)
```
minor thing, but I wonder if the stuff in parens should be `BAM Transaction ID`? Then all of the letters in the attribute name are correlated with the descriptive name - as of now we're missing the words for "am"

duglin replied:
ping @katallaxie, any thoughts on ^^ ?

katallaxie replied:
@duglin I changed it to `BAM` as we are defining the `BAM` term in the header. I think then this is consistent.

duglin commented:
@JemDay are you ok with the direction of the latest set of changes?

JemDay commented:
LGTM :-)

duglin commented:
thanks @katallaxie - and thanks for your patience.
Approved on the 5/2 call
